spec/02-integration/11-dbless/03-config_persistence.lua

```diff
@@ -0,0 +1,157 @@
+
+    local res = admin_client:post("/config",{
+      body = { config = config },
+      headers = {
+        ["Content-Type"] = "application/json"
+      }
+    })
+    assert.res_status(201, res)
+
+    assert(helpers.restart_kong({
+        database   = "off",
+    }))
+
+    proxy_client:close()
+    proxy_client = assert(helpers.proxy_client())
+
+    res = assert(proxy_client:get("/1", { headers = { host = "example1.dev" } }))
+    assert.res_status(401, res)
+    res = assert(proxy_client:get("/1000", { headers = { host = "example1.dev" } }))
+    assert.res_status(401, res)
+
+    assert.logfile().has.line("found persisted lmdb config")
+  end)
+end)
+
+describe("dbless persistence with a declarative config #off", function()
+  local admin_client, proxy_client, yaml_file
+
+  lazy_setup(function()
+    yaml_file = helpers.make_yaml_file([[
+      _format_version: "1.1"
+      services:
+      - name: my-service
+        url: https://example1.dev
+        plugins:
+        - name: key-auth
+        routes:
+        - name: my-route
+          paths:
+          - /test
+    ]])
+  end)
+
+  before_each(function()
+    assert(helpers.start_kong({
+        database   = "off",
+        declarative_config = yaml_file,
+    }))
+  
+    admin_client = assert(helpers.admin_client())
+    proxy_client = assert(helpers.proxy_client())
+
+    local res = assert(proxy_client:get("/test", { headers = { host = "example1.dev" } }))
+    assert.res_status(401, res)
+
+    local buffer = {"_format_version: '1.1'", "services:"}
+    local i = 500
+    buffer[#buffer + 1] = fmt(SERVICE_YML, i, i, i, i)
+    local config = table.concat(buffer, "\n")
+    local res = admin_client:post("/config",{
+      body = { config = config },
+      headers = {
+        ["Content-Type"] = "application/json"
+      }
+    })
+    assert.res_status(201, res)
+    res = assert(proxy_client:get("/500", { headers = { host = "example1.dev" } }))
+    assert.res_status(401, res)
+    
+    proxy_client:close()
+  end)
+    
+  after_each(function()
+    if admin_client then
+      admin_client:close()
+    end
+    if proxy_client then
+      proxy_client:close()
+    end
+    helpers.stop_kong(nil, true)
+  end)
+  
+  lazy_teardown(function()
+    os.remove(yaml_file)
+  end)
+
+  it("doesn't load the persisted lmdb config if a declarative config is set on restart", function()
+    assert(helpers.restart_kong({
+        database   = "off",
+        declarative_config = yaml_file,
+    }))
+    proxy_client = assert(helpers.proxy_client())
+    local res = assert(proxy_client:get("/test", { headers = { host = "example1.dev" } }))
+    assert.res_status(401, res)
+
+    res = assert(proxy_client:get("/500", { headers = { host = "example1.dev" } }))
+    assert.res_status(404, res) -- 404, only the declarative config is loaded
+  end)
+
+  it("doesn't load the persisted lmdb config if a declarative config is set on reload", function()
```

Comment:
For good measure, I think it would be worthwhile to have a test case that covers worker process replacement (`kill -9` an existing worker and then validate expected behavior) as well.

Reply:
@flrgh  would you consider this test a blocker? I thought that it will basically do the same thing than an initial worker start, cause the code is right in `init_worker`.

Reply:
Nah I wouldn't go as far as to say it's a blocker. I don't know this code as well as you do, so I think this is up to your discretion whether more tests are needed before merging. To clarify, there's nothing in particular about this changeset that makes me believe it will introduce any _new_ bugs. I just think our test coverage of single worker replacement is lacking overall (it's a very common blind spot I've observed in OpenResty code), so I'll almost always advocate for adding more tests when `init_worker` is concerned.

Reply:
makes sense @flrgh! I will check to add one before merging.